File: Design-Patterns-Using-CSharp/Mediator/MediatorPattern1.cs
using System;
using System.Collections.Generic;

interface IMediator
{
    void AddColleague(Colleague colleague);
    void SendMessage(string message, string colleagueId);
}

class ConcreteMediator : IMediator
{
    private Dictionary<string, Colleague> colleagues = new Dictionary<string, Colleague>();

    public void AddColleague(Colleague colleague)
    {
        colleagues[colleague.GetId()] = colleague;
    }

    public void SendMessage(string message, string colleagueId)
    {
        Console.WriteLine($"Mediator pass Message : {message}");
        colleagues[colleagueId].Receive(message);
    }
}

abstract class Colleague
{
    protected IMediator mediator;
    protected string id;

    public Colleague(IMediator mediator, string id)
    {
        this.mediator = mediator;
        this.id = id;
    }

    public abstract void Send(string message, string to);

    public abstract void Receive(string message);

    public string GetId()
    {
        return id;
    }
}

class ConcreteColleague1 : Colleague
{
    public ConcreteColleague1(IMediator mediator) : base(mediator, "First") { }

    public override void Send(string message, string to)
    {
        Console.WriteLine($"{id} Sent Message : {message}");
        mediator.SendMessage(message, to);
    }

    public override void Receive(string message)
    {
        Console.WriteLine($"{id} Received Message {message}");
    }
}

class ConcreteColleague2 : Colleague
{
    public ConcreteColleague2(IMediator mediator) : base(mediator, "Second") { }

    public override void Send(string message, string to)
    {
        Console.WriteLine($"{id} Sent Message : {message}");
        mediator.SendMessage(message, to);
    }

    public override void Receive(string message)
    {
        Console.WriteLine($"{id} Received Message {message}");
    }
}

// Client code

public class MediatorPattern1
{
    public static void Main(string[] args)
    {
        ConcreteMediator mediator = new ConcreteMediator();
        ConcreteColleague1 first = new ConcreteColleague1(mediator);
        mediator.AddColleague(first);
        ConcreteColleague2 second = new ConcreteColleague2(mediator);
        mediator.AddColleague(second);

        first.Send("Hello, World!", "Second");
    }
}

/*
First Sent Message : Hello, World!
Mediator pass Message : Hello, World!
Second Received Message Hello, World!
*/

File: Design-Patterns-Using-Go/Mediator/MediatorPattern1.go
package main

import (
	"fmt"
)

// Mediator interface
type Mediator interface {
	addColleague(colleague Colleague)
	sendMessage(message string, colleagueID string)
}

// ConcreteMediator struct
type ConcreteMediator struct {
	colleagues map[string]Colleague
}

func (m *ConcreteMediator) addColleague(colleague Colleague) {
	m.colleagues[colleague.getID()] = colleague
}

func (m *ConcreteMediator) sendMessage(message string, colleagueID string) {
	fmt.Printf("Mediator pass Message: %s\n", message)
	m.colleagues[colleagueID].receive(message)
}

// Colleague interface
type Colleague interface {
	getID() string
	send(message string, to string)
	receive(message string)
}

// ConcreteColleague1 struct
type ConcreteColleague1 struct {
	mediator Mediator
	id       string
}

func (c *ConcreteColleague1) getID() string {
	return c.id
}

func (c *ConcreteColleague1) send(message string, to string) {
	fmt.Printf("%s Sent Message: %s\n", c.id, message)
	c.mediator.sendMessage(message, to)
}

func (c *ConcreteColleague1) receive(message string) {
	fmt.Printf("%s Received Message %s\n", c.id, message)
}

// ConcreteColleague2 struct
type ConcreteColleague2 struct {
	mediator Mediator
	id       string
}

func (c *ConcreteColleague2) getID() string {
	return c.id
}

func (c *ConcreteColleague2) send(message string, to string) {
	fmt.Printf("%s Sent Message: %s\n", c.id, message)
	c.mediator.sendMessage(message, to)
}

func (c *ConcreteColleague2) receive(message string) {
	fmt.Printf("%s Received Message %s\n", c.id, message)
}

// Client code
func main() {
	mediator := &ConcreteMediator{
		colleagues: make(map[string]Colleague),
	}

	first := &ConcreteColleague1{
		mediator: mediator,
		id:       "First",
	}
	mediator.addColleague(first)

	second := &ConcreteColleague2{
		mediator: mediator,
		id:       "Second",
	}
	mediator.addColleague(second)

	first.send("Hello, World!", "Second")
}

/*
First Sent Message: Hello, World!
Mediator pass Message: Hello, World!
Second Received Message Hello, World!
*/